smolyay/surrogate.py

```diff
@@ -192,31 +192,52 @@ def gradient(self, x):
         """
         if self._coefficients is None:
             raise ValueError('Function needs training!')
-        x = numpy.array(x, copy=False, ndmin=1)
-        if x.shape != (self.dimension, ):
-            raise IndexError('Input must match dimension of domain.')
-        if not (numpy.all(x >= self._domain[:, 0])
-                and numpy.all(x <= self._domain[:, 1])):
+        input_shape = numpy.shape(x)
+        x = numpy.array(x, copy=False, ndmin=2)
+        if self.dimension == 1:
+            if len(input_shape) <= 1:
+                # the cast to 2d puts these in wrong order, so transpose
+                x = x.T
+            elif x.shape[-1] > 1:
+                x = x[..., numpy.newaxis]
+            elif input_shape == x.shape:
+                input_shape = input_shape[:-1]
+
+        if x.shape[-1] != self.dimension:
+            raise IndexError("Input must match dimension of domain")
+        if self.dimension > 1:
+            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
+                    numpy.any(x[..., i] > self.domain[i][1]) 
+                    for i in range(self.dimension))
+        else:
+            oob = (numpy.any(x < self.domain[0]) or 
+                    numpy.any(x > self.domain[1]))
+        if oob:
             raise ValueError('x must lie in domain of surrogate')
-
-        gradient = []
         # transform point into basis domain
         x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
         # evaluate the gradient
+        gradient = numpy.zeros(x_scaled.shape)
         for ni in range(self.dimension):
             value = 0
             for coeff, basis in zip(self.coefficients,
                                     self.grid.basis_functions):
                 if self.dimension > 1:
                     term = numpy.prod(
-                        [f.derivative(x) if dim == ni else f(x) for dim, (x, f)
-                         in enumerate(zip(x_scaled, basis))])
+                            [basis[dim].derivative(x_scaled[...,dim]) 
+                             if dim == ni else basis[dim](x_scaled[...,dim]) 
+                             for dim in range(len(basis))], axis=0)
                 else:
-                    term = basis.derivative(x_scaled[0])
+                    term = basis.derivative(x_scaled)
+                
                 value += coeff*term
-            gradient.append(value)
+            gradient[...,ni] = numpy.reshape(value, gradient.shape[:-1])
 
-        return gradient[0] if self.dimension == 1 else gradient
+        if input_shape == ():
```

Comment:
Per my first review:
```suggestion
        if len(input_shape) == 0:
```
The gradient is only scalar for 1d problems with a scalar input, which is the only case where the input shape has length 0. In any other case, the shape of the gradient must match the input shape so the second condition is fine.